feat: deploy response time collector

aceforeverd opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

close #1234 


* **What is the current behavior?** (You can also link to an open issue here)



* **What is the new behavior (if this is a feature change)?**

aceforeverd committed:
feat: add query_response_time statistics collector

aceforeverd committed:
feat: add deploy query time collector

aceforeverd committed:
add tests for query response time collector

aceforeverd committed:
make deploy response time basic runnable

but it is error prone
1. not safe between clusters, updating query_response_time might need a
   distributed lock
2. not thread safe neither, since it still allow getting the stats and
   mutate stats parally
3. lack of tests

aceforeverd committed:
test: enable clang thread safety analysis with absl's thread annotation

aceforeverd committed:
feat: collect and write deploy response times in nameserver

aceforeverd commented on src/tablet/tablet_impl.cc:
```diff
@@ -4863,6 +4891,14 @@ void TabletImpl::DropProcedure(RpcController* controller, const ::openmldb::api:
+    auto collector_key = absl::StrCat(db_name, ".", sp_name);
+    auto s = deploy_collector_->DeleteDeploy(collector_key);
+    if (!s.ok()) {
+        LOG(ERROR) << "[ERROR] delete deploy collector: " << s;
```
check if it is a deployment

aceforeverd commented on src/tablet/tablet_impl.cc:
```diff
@@ -1616,13 +1624,23 @@ void TabletImpl::SQLBatchRequestQuery(RpcController* ctrl, const openmldb::api::
 void TabletImpl::ProcessBatchRequestQuery(RpcController* ctrl,
                                           const openmldb::api::SQLBatchRequestQueryRequest* request,
                                           openmldb::api::SQLBatchRequestQueryResponse* response, butil::IOBuf& buf) {
+    absl::Time start = absl::Now();
```
check `Query` rpc for deployment query

dl239 commented on src/nameserver/name_server_impl.cc:
```diff
@@ -10365,5 +10373,146 @@ base::Status NameServerImpl::InitGlobalVarTable() {
+
+    // Step two: Fetch And Flush deploy stats from each tablet
+    statistics::DeployResponseTimeRowReducer reducer;
+    for (auto& kv : tablets_) {
```
holding the lock to access `tablets_`

dl239 commented on src/tablet/tablet_impl.cc:
```diff
@@ -4849,6 +4867,16 @@ void TabletImpl::CreateProcedure(RpcController* controller, const openmldb::api:
 
+    if (sp_info.type() == type::ProcedureType::kReqDeployment) {
+        auto collector_key = absl::StrCat(db_name, ".", sp_name);
+        auto s = deploy_collector_->AddDeploy(collector_key);
```
AddDeploy in `RefreshTableInfo`

aceforeverd replied:
how about lazy recover from `sp_cache_`, so in https://github.com/4paradigm/OpenMLDB/pull/1521/files#diff-eeaff339ba9a4857121028e73b5e1f32beff596c961a3a270cf028709c3c1ee9R5041, if it failed because of deploy key not found, will retry with `AddDeploy` first. The benefit is, as long as `sp_cache_` is corret:
- we don't even have to handle `create procedure` event
- when `deploy_stats` not enabled, there is no overhead for `create procedure`. though that overhead is low already

aceforeverd committed:
fix comments

aceforeverd committed:
build: fix finding protobuf on non-standard OS

use cmake based `find_package` to find cmake, which give more control
and returned variables

aceforeverd committed:
build: don't fail build if can't find rocksdb with cmake targets

aceforeverd committed:
fix compile

aceforeverd committed:
improve NameServerImpl::GetSdkConnection()

aceforeverd committed:
add test for DEPLOY RESPONSE TIME

aceforeverd committed:
fix some logs

aceforeverd committed:
refactor query_response_time

aceforeverd committed:
fix sql_cluster_avilable_test

add sanity check when trying to delete records from DEPLOY_RESPONSE_TIME
in name_server

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into feat/deploy-response-time-collector

aceforeverd commented on src/nameserver/name_server_impl.cc:
```diff
@@ -18,7 +18,15 @@
+#include "absl/time/time.h"
 #include "nameserver/system_table.h"
+#include "statistics/query_response_time/deploy_query_response_time.h"
+#include "statistics/query_response_time/query_response_time.h"
```
cleanup

aceforeverd committed:
update sql_cmd_test

dl239 commented on src/nameserver/name_server_impl.cc:
```diff
@@ -9973,12 +9983,52 @@ void NameServerImpl::DropProcedureOnTablet(const std::string& db_name, const std
+        std::lock_guard<std::mutex> lock(mu_);
+        auto it = db_sp_info_map_.find(db_name);
+        if (it != db_sp_info_map_.end()) {
+            is_deployment_procedure = it->second.find(sp_name) != it->second.end();
```
check type in procedure info to make sure it is a deployment?

aceforeverd replied:
uh right

dl239 commented on src/nameserver/name_server_impl.cc:
```diff
@@ -9973,12 +9983,52 @@ void NameServerImpl::DropProcedureOnTablet(const std::string& db_name, const std
+            is_deployment_procedure = it->second.find(sp_name) != it->second.end();
+        }
+    }
+    std::shared_ptr<TableInfo> info = std::make_shared<TableInfo>();
```
it is no need to assign value to info.

dl239 commented on src/nameserver/name_server_impl.cc:
```diff
@@ -9973,12 +9983,52 @@ void NameServerImpl::DropProcedureOnTablet(const std::string& db_name, const std
+            std::string msg;
+            while (cnt++ < TIME_DISTRIBUTION_BUCKET_COUNT - 1) {
+                auto key = absl::StrCat(deploy_name, "|", std::to_string(time));
+                if (!tb_client->Delete(info->tid(), pid, key, "", msg)) {
```
call delete by leader tablet client

dl239 commented on src/nameserver/name_server_impl.cc:
```diff
@@ -10365,5 +10415,190 @@ base::Status NameServerImpl::InitGlobalVarTable() {
+            }
+
+            ::openmldb::api::DeployStatsResponse res;
+            if (!kv.second->client_->GetAndFlushDeployStats(&res)) {
```
do not send rpc request in the lock scope

aceforeverd committed:
ressolve a few comments

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into feat/deploy-response-time-collector

zhanghaohit reviewed (approved):
LGTM
